File: caca_palavra.py
def le_matriz(matriz, m):
    for i in range(m):
        linha = input().split()
        matriz.append(linha)
    return matriz


def cria_matriz_ponto(m, n):
    matriz_ponto = []
    for i in range(m):
        linha = []
        for j in range(n):
            linha.append('.')
        matriz_ponto.append(linha)
    return matriz_ponto
    


def print_matriz(matriz, m, n):
    for i in range(m):
        for j in range(n):
            print(matriz[i][j], end=' ')
        print()


def copia_inverso(palavra):
    inverso = ''
    i = len(palavra) - 1
    while i >= 0:
        inverso += palavra[i]
        i -= 1
    return inverso


def marca_ocorrencia_horizontal(palavra, A, B, m, n, i, j):
    pos = []
    counter = 0
    tamanho = len(palavra)
    for i in range(m):
        for j in range(n):
            if A[i][j] == palavra[counter]:
                counter += 1
                pos.append([i,j])
            if counter == tamanho:
                for posicao in pos:
                    B[i]   


def search_word(A, B, word):
    directions = [[0, 1,], [1, 0], [1, 1], [-1, 0], [0, -1], [-1, -1], [-1, 1], [1, -1]]
    word_length = len(word)
    
    for row in range(len(A)):
        for col in range(len(A[0])):
            for direction in directions:
                dr, dc = direction
                end_row = row + (word_length - 1) * dr
                end_col = col + (word_length - 1) * dc
                
                if 0 <= end_row < len(A) and 0 <= end_col < len(A[0]):
                    temp_row, temp_col = row, col
                    found = True
                    
                    for char in word:
                        if A[temp_row][temp_col] != char:
                            found = False
                            break
                        temp_row += dr
                        temp_col += dc
                    
                    if found:
                        temp_row, temp_col = row, col
                        for char in word:
                            B[temp_row][temp_col] = char
                            temp_row += dr
                            temp_col += dc

dimen = input().split()
m = int(dimen[0])
n = int(dimen[1])
matriz = []

A = le_matriz(matriz, m) # matriz letra
B = cria_matriz_ponto(m, n) # matriz ponto

num_palavras = int(input())
for i in range(num_palavras):
    palavra = input()
    search_word(A, B, palavra)
    print_matriz(B, m, n)

File: caca_palavra.c
#include <stdio.h>
#include <string.h>


#define MAX_PALAVRA 15

#define MAX_ALTURA 15

#define MAX_LARGURA 10

void le_matriz(char matriz[MAX_ALTURA][MAX_LARGURA],int m,int n) {
    int i,j;
    char letra;
    for (i = 0; i < m; i++){
        for (j = 0; j < n; j++){
            scanf(" %c", &letra);
            matriz[i][j] = letra;
        }
    }
}

void cria_matriz_ponto(char matriz[MAX_ALTURA][MAX_LARGURA], int m, int n){
    int i,j;
    for (i = 0; i < m; i++){
        for (j = 0; j < n; j++){
            matriz[i][j] = '.';
        }
    }
}

void imprime_matriz(char matriz[MAX_ALTURA][MAX_LARGURA], int m, int n){
    int i,j;
    for (i = 0; i < m; i++){
        for (j = 0; j < n; j++){
            printf("%c ",matriz[i][j]);
        }
        printf("\n");
    }
}

void search_word(char A[MAX_ALTURA][MAX_LARGURA], char B[MAX_ALTURA][MAX_LARGURA], char word[], int m, int n){
    int directions[8][2] = {{0, 1}, {1, 0}, {1, 1}, {-1, 0}, {0, -1}, {-1, -1}, {-1, 1}, {1, -1}};
    int word_length = strlen(word);
    int i,j,d,k;
    for (i = 0; i < m; i++){
        for (j = 0; j < n; j++){
            for (d = 0; d < 8; d++){
                int dr = directions[d][0];
                int dc = directions[d][1];
                int end_row = (word_length - 1) * dr + i;
                int end_col = (word_length - 1) * dc + j;

                if (end_row >= 0 && end_row < m && end_col >= 0 && end_col < n){
                    int temp_row = i;
                    int temp_col = j;
                    int found = 1;

                    for (k = 0; k < word_length; k++){
                        if (A[temp_row][temp_col] != word[k]){
                            found = 0;
                            break;
                        }
                    temp_row += dr;
                    temp_col += dc;
                    }
                    if (found){
                        temp_row = i;
                        temp_col = j;
                        for (i = 0; i < word_length; i++){
                            B[temp_row][temp_col] = word[i];
                            temp_row += dr;
                            temp_col += dc;
                        }
                    }
                }
            }
        }
    }
}

int main(){
    int i;
    int m,n;
    char palavra[MAX_PALAVRA], inverso[MAX_PALAVRA];
    char A[MAX_ALTURA][MAX_LARGURA];
    char B[MAX_ALTURA][MAX_LARGURA];
    int p;

    scanf("%d %d", &m, &n); // le as dim. da matriz
    le_matriz(A, m, n);
    cria_matriz_ponto(B, m, n);

    scanf("%d", &p);
    for (i = 0; i < p; i++){
        scanf("%s", palavra);
        search_word(A, B, palavra, m, n);
    }
    imprime_matriz(B, m, n);
}